planner/core/push_down_predicate_to_table_scan.go

```diff
@@ -0,0 +1,305 @@
+	selectivity float64
+}
+
+func predicatePushDownToTableScan(sctx sessionctx.Context, plan PhysicalPlan) {
+	switch p := plan.(type) {
+	case *PhysicalSelection:
+		physicalTableScan, ok := plan.Children()[0].(*PhysicalTableScan)
+		if ok && physicalTableScan.isFullScan() && physicalTableScan.StoreType == kv.TiFlash {
+			// Only when the table scan is a full scan and the store type is TiFlash,
+			// we will try to push down predicates.
+			predicatePushDownToTableScanImpl(sctx, p, physicalTableScan)
+		} else if !ok {
+			for _, child := range plan.Children() {
+				predicatePushDownToTableScan(sctx, child)
+			}
+		}
+	case *PhysicalTableReader:
+		predicatePushDownToTableScan(sctx, p.tablePlan)
+	default:
+		for _, child := range plan.Children() {
+			predicatePushDownToTableScan(sctx, child)
+		}
+	}
+}
+
+/**
+ * @brief: This function is used to remove the empty selection in plan
+ * 		   when late materialization is enabled and all conditions are pushed down.
+ * @param: plan: the physical plan
+ */
+func removeEmptySelection(plan PhysicalPlan) {
+	if tableReader, isTableReader := plan.(*PhysicalTableReader); isTableReader && tableReader.StoreType == kv.TiFlash {
+		if selection, isSelection := tableReader.tablePlan.(*PhysicalSelection); isSelection && len(selection.Conditions) == 0 {
+			if len(tableReader.tablePlan.Children()) == 1 {
+				tableReader.tablePlan = tableReader.tablePlan.Children()[0]
+			} else {
+				logutil.BgLogger().Warn("invalid table plan", zap.String("plan", plan.ExplainID().String()))
+			}
+		}
+		removeEmptySelection(tableReader.tablePlan)
+		tableReader.TablePlans = flattenPushDownPlan(tableReader.tablePlan)
+	} else {
+		var newChildren []PhysicalPlan
+		removed := false
+		for _, child := range plan.Children() {
+			switch x := child.(type) {
+			case *PhysicalSelection:
+				if len(x.Conditions) == 0 {
+					newChildren = append(newChildren, x.children...)
+					removed = true
+				} else {
+					newChildren = append(newChildren, x)
+				}
+			default:
+				newChildren = append(newChildren, x)
+			}
+		}
+		if removed && len(newChildren) > 0 {
+			plan.SetChildren(newChildren...)
+		}
+		for _, child := range plan.Children() {
+			removeEmptySelection(child)
+		}
+	}
+}
+
+/**
+ * @brief: This function is used to transform the columns to a string of "0" and "1".
+ * @param: cols: the columns of a Expression
+ * @param: totalColumnCount: the total number of columns in the tablescan
+ * @example: the columns of tablescan are [a, b, c, d, e, f, g, h]
+ * 		 	 the expression are ["a > 1", "c > 1", "e > 1 or g > 1"]
+ *           so the columns of the expression are [a, c, e, g] and the totalColumnCount is 8
+ *           the return value is "10101010"
+ * @return: the string of "0" and "1"
+ */
+func transformColumnsToCode(cols []*expression.Column, totalColumnCount int) string {
+	code := make([]byte, totalColumnCount)
+	for _, col := range cols {
+		code[col.Index] = '1'
+	}
+	return string(code)
+}
```

Comment:
Why not use bitmask?

Reply:
Can bitmask handle 64+ columns?

Reply:
oh you want to use it to do hashing. I think it's proper way.